README.md
# [DEPRECATED] axe-cli

[![No Maintenance Intended](http://unmaintained.tech/badge.svg)](http://unmaintained.tech/)

> This repository has been deprecated. The package has been moved to [axe-core-npm](https://github.com/dequelabs/axe-core-npm/tree/develop/packages/cli). The package will be available via NPM as [`@axe-core/cli`](https://www.npmjs.com/package/@axe-core/cli).

---

[![Greenkeeper badge](https://badges.greenkeeper.io/dequelabs/axe-cli.svg)](https://greenkeeper.io/)

[![Join the chat at https://gitter.im/dequelabs/axe-core](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/dequelabs/axe-core?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Version](https://img.shields.io/npm/v/axe-cli.svg)](https://www.npmjs.com/package/axe-cli)
[![License](https://img.shields.io/npm/l/axe-cli.svg)](LICENSE)

Provides a command line interface for [aXe](https://github.com/dequelabs/axe-core) to run quick accessibility tests.

## Getting Started

Install [Node.js](https://docs.npmjs.com/getting-started/installing-node) if you haven't already. This project requires Node 6+. By default, axe-cli runs Chrome in headless mode, which requires Chrome 59 or up.

Install axe-cli globally: `npm install axe-cli -g`

Lastly, install the webdrivers of the browsers you wish to use. A webdriver is a driver for your web browsers. It allows other programs on your machine to open a browser and operate it. Current information about available webdrivers can be found at [selenium-webdriver project](https://www.npmjs.com/package/selenium-webdriver). Alternatively, you could use [Webdriver manager](https://www.npmjs.com/package/webdriver-manager)

## Usage

After installing, you can now run the `axe` command in your CLI, followed by the URL of the page you wish to test:

```
axe https://www.deque.com
```

You can run multiple pages at once, simply add more URLs to the command. Keep in mind that axe-cli is not a crawler, so if you find yourself testing dozens of pages at once, you may want to consider switching over to something like [axe-webdriverjs](https://www.npmjs.com/package/axe-webdriverjs). If you do not specify the protocol, http will be used by default:

```
axe www.deque.com, dequeuniversity.com
```

**Note:** If you are having difficulty with the color scheme, use `--no-color` to disable text styles.

## Running specific rules

You can use the `--rules` flag to set which rules you wish to run, or you can use `--tags` to tell axe to run all rules that have that specific tag. For example:

```
axe www.deque.com --rules color-contrast,html-has-lang
```

Or, to run all wcag2a rules:

```
axe www.deque.com --tags wcag2a
```

In case you want to disable some rules, you can use `--disable` followed by a list of rules. These will be skipped when analyzing the site:

```
axe www.deque.com --disable color-contrast
```

This option can be combined with either `--tags` or `--rules`.

A list of rules and what tags they have is available at: https://dequeuniversity.com/rules/worldspace/3.0/.

## Saving the results

Results can be saved as JSON data, using the `--save` and `--dir` flags. By passing a filename to `--save` you indicate how the file should be called. If no filename is passed, a default will be used. For example:

```
axe www.deque.com --save deque-site.json
```

Or:

```
axe www.deque.com --dir ./axe-results/
```

## Sending results to STDOUT

To output the test results to STDOUT, provide the `--stdout` flag. This flag has the side-effect of silencing all other logs/output (other than errors, which are written to STDERR).

To print the entire result object to your terminal, do:

```
axe --stdout www.deque.com
```

To pipe the results to a file, do:

```
axe --stdout www.deque.com > your_file.json
```

To pipe the results to a JSON-parsing program for further processing, do:

```
axe --stdout www.deque.com | jq ".[0].violations"
```

## Defining the scope of a test

If you want to only test a specific area of a page, or wish to exclude some part of a page you can do so using the `--include` and `--exclude` flags and pass it a CSS selector:

```
axe www.deque.com --include "#main" --exclude "#aside"
```

You may pass multiple selectors with a comma-delimited string. For example:

```
axe www.deque.com --include "#div1,#div2,#div3"
```

## Custom axe-core versions

Axe-cli will look for locally available versions of axe-core. If the directory from where you start axe-cli has an `axe.js` file, or has a `node_modules` directory with axe-core installed in it. Axe-cli will use this version of axe-core instead of the default version installed globally.

To specify the exact file axe-core file axe-cli should use, you can use the `--axe-source` flag (`-a` for short), with a relative or absolute path to the file.

```
axe www.deque.com --axe-source ./axe.nl.js
```

## Different browsers

Axe-cli can run in a variety of web browsers. By default axe-cli uses Chrome in headless mode. But axe-cli is equally capable of testing pages using other web browsers. **Running in another browser requires that browser's webdriver to be available on your PATH**. You can find a list of available webdrivers and how to install them at: https://seleniumhq.github.io/docs/wd.html

To run axe-cli using another browser, pass it in as the `--browser` option:

```
axe www.deque.com --browser chrome
```

Or for short:

```
axe www.deque.com -b c
```

## Custom Chrome Flags

When using the Headless Chrome browser, you may provide any number of [flags to configure how the browser functions](https://peter.sh/experiments/chromium-command-line-switches/).

Options are passed by name, without their leading `--` prefix. For example, to provide the `--no-sandbox --disable-setuid-sandbox --disable-dev-shm-usage` flags to the Chrome binary, you'd do:

```
axe --chrome-options="no-sandbox,disable-setuid-sandbox,disable-dev-shm-usage" www.deque.com
```

## CI integration

Axe-cli can be ran within the CI tooling for your project. Many tools are automatically configured to halt/fail builds when a process exits with a code of `1`.

Use the `--exit` flag, `-q` for short, to have the axe-cli process exit with a failure code `1` when any rule fails to pass.

```
axe www.deque.com --exit
```

## Timing and timeout

For debugging and managing timeouts, there are two options available. With `--timer` set, axe-cli will log how long it takes to load the page, and how long it takes to run axe-core. If you find the execution of axe-core takes too long, which can happen on very large pages, use `--timeout` to increase the time axe has to test that page:

```
axe www.cnn.com --timeout=120
```

## Delay audit to ensure page is loaded

If you find your page is not ready after axe has determined it has loaded, you can use `--load-delay` followed by a number in milliseconds. This will make axe wait that time before running the audit after the page has loaded.

```
axe www.deque.com --load-delay=2000
```

## Verbose output

To see additional information like test tool name, version and environment details, use the `--verbose` flag, `-v` for short.

```
axe www.deque.com --verbose
```

## ChromeDriver Path

If you need to test your page using an older version of Chrome, you can use `--chromedriver-path` followed by the absolute path to the desired version of the ChromeDriver executable.

```
axe www.deque.com --chromedriver-path="absolute/path/to/chromedriver"
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [3.2.1](https://github.com/dequelabs/axe-cli/compare/v3.2.0...v3.2.1) (2020-10-08)

### Bug Fixes

- drop node v6/7 support ([#137](https://github.com/dequelabs/axe-cli/issues/137)) ([00b23b9](https://github.com/dequelabs/axe-cli/commit/00b23b9))
- Enable using of --chrome-options argument ([#134](https://github.com/dequelabs/axe-cli/issues/134)) ([a0da6ba](https://github.com/dequelabs/axe-cli/commit/a0da6ba))

## [3.2.0](https://github.com/dequelabs/axe-cli/compare/v3.1.1...v3.2.0) (2019-09-30)

### Bug Fixes

- **circleci:** use ([7a8ca66](https://github.com/dequelabs/axe-cli/commit/7a8ca66))

### Features

- **webdriver:** add option ([134b299](https://github.com/dequelabs/axe-cli/commit/134b299))

### [3.1.1](https://github.com/dequelabs/axe-cli/compare/v3.1.0...v3.1.1) (2019-08-20)

### Bug Fixes

- Update Chromedriver dependency [#103](https://github.com/dequelabs/axe-cli/issues/103) [#104](https://github.com/dequelabs/axe-cli/issues/104)

## [3.1.0](https://github.com/dequelabs/axe-cli/compare/v3.0.0...v3.1.0) (2019-07-15)

### Bug Fixes

- correct scope documentation ([#80](https://github.com/dequelabs/axe-cli/issues/80)) ([81b4312](https://github.com/dequelabs/axe-cli/commit/81b4312)), closes [#75](https://github.com/dequelabs/axe-cli/issues/75)

### Features

- add `--chrome-options` flag ([#81](https://github.com/dequelabs/axe-cli/issues/81)) ([6214bcb](https://github.com/dequelabs/axe-cli/commit/6214bcb)), closes [#65](https://github.com/dequelabs/axe-cli/issues/65)
- add `--stdout` flag ([#83](https://github.com/dequelabs/axe-cli/issues/83)) ([06328bf](https://github.com/dequelabs/axe-cli/commit/06328bf)), closes [#15](https://github.com/dequelabs/axe-cli/issues/15)
- add meta data to cli output ([#94](https://github.com/dequelabs/axe-cli/issues/94)) ([7ee59e9](https://github.com/dequelabs/axe-cli/commit/7ee59e9))

### Tests

- make "ready class" test more forgiving ([#74](https://github.com/dequelabs/axe-cli/issues/74)) ([fc2b595](https://github.com/dequelabs/axe-cli/commit/fc2b595))

<a name="3.0.0"></a>

# [3.0.0](https://github.com/dequelabs/axe-cli/compare/v2.1.0-alpha.1...v3.0.0) (2018-03-28)

### Features

- Update to [axe-core 3.0.0](https://github.com/dequelabs/axe-core/releases/tag/v3.0.0)
- Add --load-delay option to delay audit after page loads ([#53](https://github.com/dequelabs/axe-cli/issues/53)) ([c0659a8](https://github.com/dequelabs/axe-cli/commit/c0659a8))
- Upgrade chromedriver to support Chrome 65 ([e4d4bd1](https://github.com/dequelabs/axe-cli/commit/e4d4bd1))

<a name="2.1.0-alpha.1"></a>

# [2.1.0-alpha.1](https://github.com/dequelabs/axe-cli/compare/v2.1.0-alpha.0...v2.1.0-alpha.1) (2018-02-21)

### Features

- Support aXe-core 3.0 Shadow DOM selectors ([#49](https://github.com/dequelabs/axe-cli/issues/49)) ([790b421](https://github.com/dequelabs/axe-cli/commit/790b421))

<a name="2.1.0-alpha.0"></a>

# [2.1.0-alpha.0](https://github.com/dequelabs/axe-cli/compare/v2.0.0...v2.1.0-alpha.0) (2018-02-20)

### Bug Fixes

- Security vulnerability in hoek package ([#50](https://github.com/dequelabs/axe-cli/issues/50)) ([81695ad](https://github.com/dequelabs/axe-cli/commit/81695ad))

### Features

- Upgrade axe-core to 3.0.0-beta.1
- Upgrade axe-webdriverjs to 2.0.0-alpha.1

<a name="2.0.0"></a>

## [2.0.0](https://github.com/dequelabs/axe-cli/compare/v1.3.1...v2.0.0) (2017-12-19)

### Features

- Use chrome-headless as default browser replacing PhantomJS ([1ae8e12](https://github.com/dequelabs/axe-cli/commit/1ae8e12))

### BREAKING CHANGES

- PhantomJS is no longer maintained. We will be
  replacing it with headless Chrome

<a name="1.3.1"></a>

## [1.3.1](https://github.com/dequelabs/axe-cli/compare/v1.3.0...v1.3.1) (2017-12-19)

### Features

- Add axe-core 2.6.0

<a name="1.3.0"></a>

# [1.3.0](https://github.com/dequelabs/axe-cli/compare/v1.1.1...v1.3.0) (2017-11-17)

### Bug Fixes

- package.json & .snyk to reduce vulnerabilities ([#39](https://github.com/dequelabs/axe-cli/issues/39)) ([9b20eef](https://github.com/dequelabs/axe-cli/commit/9b20eef))

### Features

- Add flag that enables supplying a list of rules to be skipped during the analysis ([d22903d](https://github.com/dequelabs/axe-cli/commit/d22903d))
- Allow running from file:// and ftp(s):// ([#41](https://github.com/dequelabs/axe-cli/issues/41)) ([aa3d937](https://github.com/dequelabs/axe-cli/commit/aa3d937))
- Link to DeqeuU courses/testingmethods ([#38](https://github.com/dequelabs/axe-cli/issues/38)) ([8c0e661](https://github.com/dequelabs/axe-cli/commit/8c0e661))

<a name="1.2.0"></a>

# [1.2.0](https://github.com/dequelabs/axe-cli/compare/1.0.2...1.2.0) (2017-10-31)

### Features

- Allow running from file:// and ftp(s):// ([#41](https://github.com/dequelabs/axe-cli/issues/41)) ([aa3d937](https://github.com/dequelabs/axe-cli/commit/aa3d937))
- Link to DeqeuU courses/testingmethods ([#38](https://github.com/dequelabs/axe-cli/issues/38)) ([8c0e661](https://github.com/dequelabs/axe-cli/commit/8c0e661))
- support exit codes ([e14e2d5](https://github.com/dequelabs/axe-cli/commit/e14e2d5)), closes [#20](https://github.com/dequelabs/axe-cli/issues/20) [#22](https://github.com/dequelabs/axe-cli/issues/22)

<a name="1.1.1"></a>

## [1.1.1](https://github.com/dequelabs/axe-cli/compare/1.0.3...1.1.1) (2017-09-20)

### New Features

- feat: Add --timeout and --timer options ([6d4d14f](https://github.com/dequelabs/axe-cli/commit/6d4d14f80e63bef2d54b3704a818a8ca8b1bb0e3))
- chore: upgrade axe-core to 2.4.1, axe-webdriverjs to 1.1.5 ([933f1fd](https://github.com/dequelabs/axe-cli/commit/933f1fdb60b06c6fbbcf6d77763dd334d4df8d73))

### Bug Fixes

- doc: Changed non-working promo url for courses to use a working url ([ca7361e](https://github.com/dequelabs/axe-cli/commit/ca7361e653ccb8f3a0138d0dc5f800ff09136351))


test/webdriver.js
'use strict';

const assert = require('chai').assert;
const { startDriver, stopDriver } = require('../lib/webdriver');
const chromedriver = require('chromedriver');
const chrome = require('selenium-webdriver/chrome');
const path = require('path');

describe('startDriver', () => {
	let config, browser;
	beforeEach(() => {
		browser = 'chrome-headless';
		config = {
			get browser() {
				return browser;
			}
		};
	});

	afterEach(async () => {
		stopDriver(config);
		const service = chrome.getDefaultService();
		if (service.isRunning()) {
			await service.stop();

			// An unfortunately hacky way to clean up
			// the service. Stop will shut it down,
			// but it doesn't reset the local state
			service.address_ = null;
			chrome.setDefaultService(null);
		}
	});

	it('creates a driver', async () => {
		await startDriver(config);

		assert.isObject(config.driver);
		assert.isFunction(config.driver.manage);
	});

	xit('sets the config.browser as the browser', done => {
		browser = 'chrome';
		startDriver(config)
			.then(config => config.driver.getCapabilities())
			.then(capabilities => {
				assert.equal(capabilities.get('browserName'), browser);
			})
			.then(done, done);
	});

	it('sets the browser as chrome with chrome-headless', async () => {
		browser = 'chrome-headless';
		await startDriver(config);
		const capabilities = await config.driver.getCapabilities();

		assert.equal(capabilities.get('browserName'), 'chrome');
	});

	it('uses the chromedriver path with chrome-headless', async () => {
		browser = 'chrome-headless';
		await startDriver(config);
		const service = chrome.getDefaultService();

		assert.equal(service.executable_, chromedriver.path);
	});

	it('uses the passed in chromedriver path with chrome-headless', async () => {
		browser = 'chrome-headless';
		config.chromedriverPath = path.relative(process.cwd(), chromedriver.path);
		await startDriver(config);
		const service = chrome.getDefaultService();

		assert.notEqual(config.chromedriverPath, chromedriver.path);
		assert.equal(service.executable_, config.chromedriverPath);
	});

	it('sets the --headless flag with chrome-headless', async () => {
		browser = 'chrome-headless';
		const { builder } = await startDriver(config);
		const capabilities = await builder.getCapabilities();
		const chromeOptions = capabilities.get('chromeOptions');

		assert.isObject(chromeOptions);
		assert.isTrue(chromeOptions.args.includes('--headless'));
	});

	it('sets the --no-sandbox flag with chrome-headless', async () => {
		browser = 'chrome-headless';
		const { builder } = await startDriver(config);
		const capabilities = await builder.getCapabilities();
		const chromeOptions = capabilities.get('chromeOptions');

		assert.isObject(chromeOptions);
		assert.isTrue(chromeOptions.args.includes('--no-sandbox'));
	});

	it('sets the --chrome-options flag with custom args', async () => {
		browser = 'chrome-headless';
		config.chromeOptions = ['window-size=1024,768'];
		const { builder } = await startDriver(config);
		const capabilities = await builder.getCapabilities();
		const chromeOptions = capabilities.get('chromeOptions');

		assert.isArray(chromeOptions.args);
		assert.deepEqual(chromeOptions.args, [
			'--headless',
			'--no-sandbox',
			'window-size=1024,768'
		]);
	});
});

describe('stopDriver', () => {
	it('calls browser.quit', () => {
		let called = 0;
		stopDriver({
			browser: 'chrome-headless',
			driver: { quit: () => called++ }
		});
		assert.equal(called, 1);
	});
});


test/integrations.js
'use strict';

var assert = require('chai').assert;
var chrome = require('selenium-webdriver/chrome');
var http = require('http');
var nodeStatic = require('node-static');
var axeTestUrls = require('../lib/axe-test-urls');
var { startDriver, stopDriver } = require('../lib/webdriver');

describe('integrations', function() {
	var program, urls, server;

	before(function() {
		// Start a server
		var file = new nodeStatic.Server('.');
		server = http.createServer(function(request, response) {
			request
				.addListener('end', function() {
					file.serve(request, response);
				})
				.resume();
		});
		server.listen(8182);
	});

	after(function() {
		server.close();
	});

	beforeEach(async function() {
		program = {
			browser: 'chrome-headless'
		};
		await startDriver(program);
		urls = ['http://localhost:8182/test/testpage.html'];
	});

	afterEach(async () => {
		stopDriver(program);

		var service = chrome.getDefaultService();
		if (service.isRunning()) {
			await service.stop();

			// An unfortunately hacky way to clean up
			// the service. Stop will shut it down,
			// but it doesn't reset the local state
			service.address_ = null;
			chrome.setDefaultService(null);
		}
	});

	it('finds results in light and shadow DOM', async () => {
		var listResult;
		await axeTestUrls(urls, program, {
			onTestComplete: function(results) {
				assert.containsAllKeys(results, [
					'testEngine',
					'testEnvironment',
					'testRunner'
				]);
				listResult = results.violations.find(result => result.id === 'list');
				assert.lengthOf(listResult.nodes, 2);
				assert.deepEqual(listResult.nodes[0].target, ['#list']);
				assert.deepEqual(listResult.nodes[1].target, [
					['#shadow-root', '#shadow-list']
				]);
			}
		});

		assert.isDefined(listResult);
	});
});


test/axe-test-urls.js
'use strict';

const assert = require('chai').assert;
const testPages = require('../lib/axe-test-urls');

describe('testPages', function() {
	let config, mockDriver;

	beforeEach(() => {
		mockDriver = {
			get: async arg => arg,
			executeAsyncScript: async arg => arg,
			executeScript: async arg => arg,
			wait: async arg => arg,
			switchTo: () => ({ defaultContent: () => {} }),
			findElements: async () => [],
			quit: async arg => arg
		};
		config = { driver: mockDriver };
	});

	it('return a promise', () => {
		assert.instanceOf(testPages([], config, {}), Promise);
	});

	it('calls driver.get() for each URL', async () => {
		const urlsCalled = [];
		const urls = ['http://foo', 'http://bar', 'http://baz'];

		mockDriver.get = async url => {
			urlsCalled.push(url);
			return url;
		};

		await testPages(urls, config, {});

		assert.deepEqual(urlsCalled, urls);
	});

	it('waits until the document is ready to have a className added', async () => {
		const asyncScripts = [];
		let waitCalls = 0;

		mockDriver.executeAsyncScript = async script => {
			asyncScripts.push(script);
			return script;
		};
		mockDriver.wait = async script => {
			waitCalls++;
			return script;
		};

		await testPages(['http://foo'], config, {});

		assert.equal(asyncScripts.length, 2);
		const [script] = asyncScripts;
		assert.match(
			script,
			/script\.innerHTML\s*=[\s\S]*['"]document\.documentElement\.classList\.add\(['"]deque-axe-is-ready/
		);

		assert.equal(waitCalls, 1);
	});

	it('injects axe into the page', async () => {
		const scripts = [];
		config.axeSource = 'axe="hi, I am axe"';
		mockDriver.executeScript = async script => {
			scripts.push(script);
			return script;
		};

		await testPages(['http://foo'], config, {});
		assert.include(scripts[0].toString(), config.axeSource);
	});

	it('runs axe once the page is loaded', async () => {
		const asyncScripts = [];
		mockDriver.executeAsyncScript = async script => {
			asyncScripts.push(script);
			return script;
		};

		await testPages(['http://foo'], config, {});

		assert.isDefined(
			asyncScripts
				.map(script => script.toString())
				.find(script => script.match(/(axe\.run)|(axe\.a11yCheck)/))
		);
	});
});


lib/cli.js
'use strict'

const program = require('commander')
const colors = require('colors')
const link = colors.underline.blue
const error = colors.red.bold

const version = require('../package.json').version
const axeTestUrls = require('./axe-test-urls')
const saveOutcome = require('./save-outcome')
const utils = require('./utils')

module.exports = function axeCliRunner (process, console) {
  program.version(version)
    .usage('<url...> [options]')
    .option('-i, --include <list>', 'CSS selector of included elements, comma separated', utils.splitList)
    .option('-e, --exclude <list>', 'CSS selector of included elements, comma separated', utils.splitList)
    .option('-r, --rules <list>', 'IDs of rules to run, comma separated', utils.splitList)
    .option('-t, --tags <list>', 'Tags of rules to run, comma separated', utils.splitList)
    .option('-b, --browser [browser-name]', 'Which browser to run (Webdriver required)')
    .option('-s, --save [filename]', 'Save the output as a JSON file. Filename is optional')
    .option('-d, --dir <path>', 'Output directory')
    .option('-a, --axe-source <path>', 'Path to axe.js file')
    .option('-q, --exit', 'Exit with `1` failure code if any a11y tests fail')
    .option('--timeout <n>', 'Set how much time (second) axe has to run (default: 90)', 90)
    .option('--timer', 'Log the time it takes to run')
    .option('--show-errors', 'Display the full error stack')
    // TODO: Replace this with a reporter option, this required adding
    // a reporter option to axe-webdriverjs
    .option('--no-reporter', 'Turn the CLI reporter off')
    // .option('-c, --config <file>', 'Path to custom axe configuration')
    .parse(process.argv)

  program.browser = utils.parseBrowser(program.browser)
  program.axeSource = (program.axeSource)

  if (!program.axeSource) {
    const message = 'Unable to find the axe-core source file.'
    console.log(error(message))
    return
  }

  // Try to match the version of axe that's used
  const axeVersion = utils.getAxeVersion(program.axeSource)

  // Setup axe with the appropriate config
  console.log(colors.bold('Running axe-core ' + axeVersion + ' in ' + program.browser))

  // Make valid URLs of all pages
  const urls = program.args.map(utils.parseUrl)
  if (urls.length === 0) {
    console.log(error(
      'No url was specified. Check `axe -h` for help\n'
    ))
    process.exitCode = 1
    return
  }

  return testUrls(urls, program, console)
}

function testUrls (process, console, urls, program) {
  let cliReporter
  if (program.reporter === false) {
    cliReporter = function () {}
  } else {
    cliReporter = function (...args) {
      console.log(...args)
    }
  }

  // Run axe inside the pages
  axeTestUrls(urls, program, {
    /**
     * Inform the user what page is tested
     */
    onTestStart: function (url) {
      console.log(
        colors.bold('\nTesting ' + link(url)) +
        ' ... please wait, this may take a minute.'
      )
      if (program.timer) {
        console.time('Total test time')
      }
    },

    /**
     * Put the result in the console
     */
    onTestComplete: function logResults (results) {
      const violations = results.violations
      if (violations.length === 0) {
        cliReporter(colors.green('  0 violations found!'))
        return
      }

      const issueCount = violations.reduce((count, violation) => {
        const errorMessage = error('  Violation of %j with %d occurrences!\n') +
          '    %s. Correct invalid elements at:\n' +
          '%s' +
          '    For details, see: %s'
        cliReporter('\n' + errorMessage,
          violation.id,
          violation.nodes.length,
          violation.description,
          violation.nodes.map(node =>
            '     - ' + node.target + '\n'
          ).join(''),
          link(violation.helpUrl.split('?')[0])
        )
        return count + violation.nodes.length
      }, 0)

      cliReporter(error('\n%d Accessibility issues detected.'), issueCount)

      if (program.exit) {
        process.exitCode = 1
      }
    }
  }).then(function (outcome) {
    console.log('')
    if (program.timer) {
      console.timeEnd('Total test time')
    }
    // All results are in, quit the browser, and give a final report
    if (outcome.length > 1) {
      console.log(colors.bold.underline(
        'Testing complete of %d pages\n'), outcome.length)
    } else if (program.timer) {
      console.log('')
    }

    // Save the outcome
    if (program.save || program.dir) {
      return saveOutcome(outcome, program.save, program.dir)
        .then(fileName => {
          console.log('Saved file at', fileName, '\n')
        }).catch(err => {
          console.log(error('Unable to save file!\n') + err)
          process.exitCode = 1
          return Promise.resolve()
        })
    } else {
      return Promise.resolve()
    }
  }).then(() => {
    // Give a notification that 0 issues in axe doesn't mean perfect a11y
    console.log(colors.italic(
      'Please note that only 20% to 50% of all accessibility ' +
      'issues can automatically be detected. \nManual testing is ' +
      'always required. For more information see:\n%s\n'
    ), link(
      'https://dequeuniversity.com/curriculum/courses/testing'
    ))
  }).catch((e) => {
    console.log(' ')
    if (!program['show-errors']) {
      console.log(error('An error occurred while testing this page.'))
    } else {
      console.log(
        error('Error: %j \n $s'),
        e.message,
        e.stack
      )
    }

    console.log('Please report the problem to: ' +
      link('https://github.com/dequelabs/axe-cli/issues/') + '\n')
    process.exit(1)
  })
}


lib/utils.js
module.exports.parseUrl = function parseUrl(url) {
	if (!/[a-z]+:\/\//.test(url)) {
		return 'http://' + url;
	}
	return url;
};

module.exports.parseBrowser = function parseBrowser(browser) {
	if (!browser) {
		return 'chrome-headless';
	}

	const l = browser.length;
	switch (browser.toLowerCase()) {
		case 'ff':
		case 'firefox'.substr(0, l):
		case 'gecko'.substr(0, l):
		case 'marionette'.substr(0, l):
			return 'firefox';

		case 'chrome'.substr(0, l):
			return 'chrome';

		case 'ie':
		case 'explorer'.substr(0, l):
		case 'internetexplorer'.substr(0, l):
		case 'internet_explorer'.substr(0, l):
		case 'internet-explorer'.substr(0, l):
			return 'ie';

		case 'safari'.substr(0, l):
			return 'safari';

		case 'edge'.substr(0, l):
		case 'microsoftedge'.substr(0, l):
			return 'MicrosoftEdge';

		default:
			throw new Error('Unknown browser ' + browser);
	}
};

module.exports.getAxeSource = function getAxeSource(axePath) {
	const path = require('path');
	const fs = require('fs');
	// Abort if axePath should exist, and it isn't
	if (axePath && !fs.existsSync(axePath)) {
		return;
		// Look for axe in CWD
	} else if (!axePath) {
		axePath = path.join(process.cwd(), './axe.js');
	}

	if (!fs.existsSync(axePath)) {
		// Look for axe in CDW ./node_modules
		axePath = path.join(process.cwd(), './node_modules/axe-core/axe.js');
	}
	if (!fs.existsSync(axePath)) {
		// if all else fails, use the locally installed axe
		axePath = path.join(__dirname, '../node_modules/axe-core/axe.js');
	}

	return fs.readFileSync(axePath, 'utf8');
};

module.exports.getAxeVersion = function getAxeVersion(source) {
	const match = source.match(/\.version\s*=\s'([^']+)'/);
	return match ? match[1] : 'unknown version';
};

module.exports.splitList = function(val) {
	return val.split(/[,;]/).map(str => str.trim());
};

module.exports.selectorToString = function(selectors, separator) {
	separator = separator || ' ';
	return selectors
		.reduce((prev, curr) => prev.concat(curr), [])
		.join(separator);
};


index.js
'use strict';

const program = require('commander');
const colors = require('colors');
const link = colors.underline.blue;
const error = colors.red.bold;
const version = require('./package.json').version;
const axeTestUrls = require('./lib/axe-test-urls');
const saveOutcome = require('./lib/save-outcome');
const utils = require('./lib/utils');

program
	.version(version)
	.usage('<url...> [options]')
	.option(
		'-i, --include <list>',
		'CSS selector of included elements, comma separated',
		utils.splitList
	)
	.option(
		'-e, --exclude <list>',
		'CSS selector of included elements, comma separated',
		utils.splitList
	)
	.option(
		'-r, --rules <list>',
		'IDs of rules to run, comma separated',
		utils.splitList
	)
	.option(
		'-t, --tags <list>',
		'Tags of rules to run, comma separated',
		utils.splitList
	)
	.option(
		'-l, --disable <list>',
		'IDs of rules to disable, comma separated',
		utils.splitList
	)
	.option(
		'-b, --browser [browser-name]',
		'Which browser to run (Webdriver required)'
	)
	.option(
		'-s, --save [filename]',
		'Save the output as a JSON file. Filename is optional'
	)
	.option(
		'-j, --stdout',
		'Output results to STDOUT and silence all other output'
	)
	.option('-d, --dir <path>', 'Output directory')
	.option('-a, --axe-source <path>', 'Path to axe.js file')
	.option('-q, --exit', 'Exit with `1` failure code if any a11y tests fail')
	.option(
		'--load-delay <n>',
		'Set how much time (milliseconds) axe will wait after page load before running the audit (default: 0)',
		0
	)
	.option(
		'--timeout <n>',
		'Set how much time (seconds) axe has to run (default: 90)',
		90
	)
	.option('--timer', 'Log the time it takes to run')
	.option('--show-errors', 'Display the full error stack')
	// TODO: Replace this with a reporter option, this required adding
	// a reporter option to axe-webdriverjs
	.option('--no-reporter', 'Turn the CLI reporter off')
	.option(
		'--chrome-options [options]',
		'Options to provide to headless Chrome',
		utils.splitList
	)
	.option(
		'-v, --verbose',
		'Output metadata like test tool name, version and environment'
	)
	.option(
		'--chromedriver-path <path>',
		'Absolute path to the desired chromedriver executable'
	)
	// .option('-c, --config <file>', 'Path to custom axe configuration')
	.parse(process.argv);

const silentMode = !!program.stdout;

program.browser = utils.parseBrowser(program.browser);
program.axeSource = utils.getAxeSource(program.axeSource);

if (!program.axeSource) {
	console.error(error('Unable to find the axe-core source file.'));
	return;
}

if (program.chromeOptions) {
	if (program.browser !== 'chrome-headless') {
		console.error(
			error('You may only provide --chrome-options when using headless chrome')
		);
		process.exit(2);
	}

	program.chromeOptions = program.chromeOptions.map(option => `--${option}`);
}

let cliReporter;
if (program.reporter === false || silentMode) {
	cliReporter = function() {};
} else {
	cliReporter = function(...args) {
		console.log(...args);
	};
}

// Try to match the version of axe that's used
const axeVersion = utils.getAxeVersion(program.axeSource);

if (!silentMode) {
	// Setup axe with the appropriate config
	console.log(
		colors.bold('Running axe-core ' + axeVersion + ' in ' + program.browser)
	);
}

// Make valid URLs of all pages
const urls = program.args.map(utils.parseUrl);

if (urls.length === 0) {
	console.error(error('No url was specified. Check `axe -h` for help\n'));
	process.exitCode = 1;
	return;
}

// Run axe inside the pages
axeTestUrls(urls, program, {
	/**
	 * Inform the user what page is tested
	 */
	onTestStart: function(url) {
		if (silentMode) {
			return;
		}

		console.log(
			colors.bold('\nTesting ' + link(url)) +
				' ... please wait, this may take a minute.'
		);
		if (program.timer) {
			console.time('Total test time');
		}
	},

	/**
	 * Put the result in the console
	 */
	onTestComplete: function logResults(results) {
		const { violations, testEngine, testEnvironment, testRunner } = results;

		if (violations.length === 0) {
			cliReporter(colors.green('  0 violations found!'));
			return;
		}

		const issueCount = violations.reduce((count, violation) => {
			cliReporter(
				'\n' +
					error('  Violation of %j with %d occurrences!\n') +
					'    %s. Correct invalid elements at:\n' +
					violation.nodes
						.map(node => '     - ' + utils.selectorToString(node.target) + '\n')
						.join('') +
					'    For details, see: %s',
				violation.id,
				violation.nodes.length,
				violation.description,
				link(violation.helpUrl.split('?')[0])
			);
			return count + violation.nodes.length;
		}, 0);

		cliReporter(error('\n%d Accessibility issues detected.'), issueCount);

		if (program.verbose) {
			const metadata = {
				'Test Runner': testRunner,
				'Test Engine': testEngine,
				'Test Environment': testEnvironment
			};
			cliReporter(`\n${JSON.stringify(metadata, null, 2)}`);
		}

		if (program.exit) {
			process.exitCode = 1;
		}
	}
})
	.then(function(outcome) {
		if (silentMode) {
			process.stdout.write(JSON.stringify(outcome, null, 2));
			return;
		}

		console.log('');
		if (program.timer) {
			console.timeEnd('Total test time');
		}
		// All results are in, quit the browser, and give a final report
		if (outcome.length > 1) {
			console.log(
				colors.bold.underline('Testing complete of %d pages\n'),
				outcome.length
			);
		} else if (program.timer) {
			console.log('');
		}

		// Save the outcome
		if (program.save || program.dir) {
			return saveOutcome(outcome, program.save, program.dir)
				.then(fileName => {
					console.log('Saved file at', fileName, '\n');
				})
				.catch(err => {
					console.error(error('Unable to save file!\n') + err);
					process.exitCode = 1;
					return Promise.resolve();
				});
		} else {
			return Promise.resolve();
		}
	})
	.then(() => {
		if (silentMode) {
			return;
		}
		// Give a notification that 0 issues in axe doesn't mean perfect a11y
		console.log(
			colors.italic(
				'Please note that only 20% to 50% of all accessibility ' +
					'issues can automatically be detected. \nManual testing is ' +
					'always required. For more information see:\n%s\n'
			),
			link('https://dequeuniversity.com/curriculum/courses/testingmethods')
		);
	})
	.catch(e => {
		console.error(' ');
		if (!program['show-errors']) {
			console.error(error('An error occurred while testing this page.'));
		} else {
			console.error(error('Error: %j \n $s'), e.message, e.stack);
		}

		console.error(
			'Please report the problem to: ' +
				link('https://github.com/dequelabs/axe-cli/issues/') +
				'\n'
		);
		process.exit(1);
	});
